lib/src/main/java/com/squareup/cycler/Update.kt

```diff
@@ -120,49 +166,68 @@ class Update<I : Any>(private val oldRecyclerData: RecyclerData<I>) {
   internal fun generateUpdateWork(itemComparator: ItemComparator<I>?): UpdateWork {
 
     val extraItemChanged = oldRecyclerData.extraItem != extraItem
-    val refreshAllNeeded = dataReplaced && itemComparator == null
 
     val asyncWork = mutableListOf<() -> Unit>()
     val notifications = mutableListOf<(Adapter<*>) -> Unit>()
 
-    if (refreshAllNeeded) {
-      @SuppressLint("NotifyDataSetChanged")
-      notifications += { adapter -> adapter.notifyDataSetChanged() }
-    } else {
+    if (extraItemChanged) {
+      // We notify first the extraItem change, so we use its position in oldRecyclerData.
+      notifications += notifyChangesExtraItem()
+    }
 
-      if (extraItemChanged) {
-        // We notify first the extraItem change, so we use its position in oldRecyclerData.
-        notifications += notifyChangesExtraItem()
+    when {
+      dataPopulated -> {
+        val work = generateUpdateLambdas(populateMode, itemComparator)
+        work.first?.let { asyncWork.add(it) }
+        notifications.add(work.second)
       }
-
-      when {
-        dataReplaced -> {
-          // refreshAllNeeded == false => itemComparator != null. We are going async.
-          lateinit var diffResult: DiffResult
-          asyncWork += {
-            diffResult = calculateDataChanges(itemComparator!!)
-          }
-          notifications += { adapter -> diffResult.dispatchUpdatesTo(adapter) }
-        }
-        dataAdded -> {
-          notifications += { adapter ->
-            val positionAt = oldRecyclerData.data.size
-            val count = addedChunks.asSequence()
-              .map(List<I>::size)
-              .sum()
-            adapter.notifyItemRangeInserted(positionAt, count)
-          }
+      dataReplaced -> {
+        val work = generateUpdateLambdas(replaceMode, itemComparator)
+        work.first?.let { asyncWork.add(it) }
+        notifications.add(work.second)
+      }
```

Comment:
I'd change this:

1) Solve type of update here:
- generateUpdateLambdas only do the diffing case (maybe call it generateDiffWork)
- solve here if it's a diffing situation or not (as we solve here if it's an adding thing or not)

2) That will help you know if it's a full update (notificationDatasetChanged) or not.
You should _not_ add notifyChangesExtraItem if it's a full data set change!

3) Have an extension function on Pair<lambda?,lambda> to make it sync:
- When you are diffing you just call generateDiffWork that returns "diffWork, and notifications"
- If updateMode is Sync you call Pair<...>.toSync()
- It will change itself to be sync

```
fun Pair<(() -> Unit)?, (Adapter) -> Unit>.toSync() =
  Pair(
    first = null, // no async work
    second = { adapter ->
      first?.invoke() // do the diffing
      second.invoke(adapter) // call adapter
    }
  )
```

Reply:
I'll mark the places where these changes apply.